lep/DRAFT/012_Pro_version_personal_server.md

```diff
@@ -0,0 +1,160 @@
+
+<p align="center">
+<img src="images/lep-012-cloud.png" alt="Payment Processing"/>
+</p>
+
+#### Redis Pro Servers queue and management
+
+If the payment was successful, the charge has been registered and we can safely launch the server and provide the server and the client with the necessary information for secure proxying.  Thus, the Lantern Pro Server needs to append a new server request in a queue in Redis **(step 1)**.  This will allow *Cloudmaster* to pick up this job and execute the requests in the best datacenter for the Lantern client **(step 2)**.  The Redis DB is effectivelly serving as a message queue between the Pro Server and the Cloudmaster, just as what happens with the Config Server and Cloudmaster for the regular service.  Cloudmaster will need watch for new Pro Servers requests, and launch as necessary.  This is currently done for regular LCSs, so it could extend the functionality or be based on it.
+
+#### Automatically configuring the client and server pairing
+
+There is some information that needs to be placed at the newly launched server, and some shared among the client and server.  We will need:
+
+* The access token (in both the client and server)
+* The server certificate (in both the client and server)
+* The server static IP (in the client)
+
+First, the server should be automatically configured with this information by the Cloudmaster.  It then needs to notify the Pro Server **(step 3)**, confirming successfull launch of the server and providing it with the information for the user.  Once this confirmation is received by the Pro Server, it will register the user service, and send an email to the user with the config file **(step 4)**.
+
+The *file-in-email* solution is proposed in order to avoid these potential issues:
+
+* The  differentiated method for purchasing Pro for one Lantern instance and several.  In other words, if you want to share the token, you inevitably need to introduce a different step for configuration.  The file can be directly shared to other Lantern instances or users and the sharing problem would be solved.  This email will contain also a direct link to the local UI to easily open the file attached and make the configuration automatic.
+
+* The time taken to spawn a new server until it becomes functional.  It's more robust to decouple the UI from the events in the Lantern Cloud.
+
+
+When the Lantern client gets all the information it requires to pair with the server, it can proceed to override config.  It is suggested that the regular free Lantern config is left as a fallback, in case there is an issue with the Pro Server, giving time to the Lantern team to investigate.
+
+
+## Sharing across devices and users
+
+These are some less critical but relevant options suggested for *Lantern Pro*:
+
+* Share token (share the Lantern Pro Server with other instances of Lantern or users).
+
+* Regenerate token (for security reasons, in case of unwanted share)
```

Comment:
Good point. I think that's possible (and was the first approach), but it could be done as a second pass, as an improvement, because it only solves the config distribution problem for the instance issuing the payment, as you mention. It is also less robust, some method for retry should be put in place for the general case that the connection is interrupted at any stage. That's why I think the email way is a robust method to be based on, and then add UX improvements on top of it.

This could work with a websocket established with the pro-server, so it can push the data through this protocol as soon as it is available. It's a cool thing to do actually :)) We can discuss its relevance for this first implementation tomorrow!